clarity/src/vm/variables.rs

```diff
@@ -92,8 +100,18 @@ pub fn lookup_reserved_variable(
             }
             NativeVariables::BlockHeight => {
                 runtime_cost(ClarityCostFunction::FetchVar, env, 1)?;
-                let block_height = env.global_context.database.get_current_block_height();
-                Ok(Some(Value::UInt(block_height as u128)))
+                // In epoch 2.x, the `block-height` keyword returns the block height, but
+                // beginning in epoch 3, it returns the tenure height instead in order to
+                // maintain a similar pace to the block  height advancement pre-Nakamoto. This
+                // keyword is removed in Clarity 3 to avoid confusion and replaced with
+                // `stacks-block-height` and `tenure-height`.
+                if env.global_context.epoch_id < StacksEpochId::Epoch30 {
+                    let block_height = env.global_context.database.get_current_block_height();
+                    Ok(Some(Value::UInt(block_height as u128)))
+                } else {
+                    let tenure_height = env.global_context.database.get_tenure_height()?;
```

Comment:
I had to read this a couple times before it made sense.  The `block-height` variable isn't going away _per se_; it's just not recognized as a global in Clarity 3.  But, it still has to have semantic meaning in both Clarity 1 and Clarity 2 contracts while in epoch 3.0.  This code is meant to handle this case specifically -- we evaluate `block-height` as the tenure height when in epoch 3.0.  This comment might need some massaging. 

Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?

Reply:
Yes, your explanation is correct and is what I was trying to get across with that comment. Does this sound better?

```rust
                // In epoch 2.x, the `block-height` keyword returns the Stacks block height.
                // For Clarity 1 and Clarity 2 contracts executing in epoch 3, `block-height`
                // is equal to the tenure height instead of the Stacks block height. This change
                // is made to maintain a similar pace at which this value increments (e.g. for use
                // as an expiration). In Clarity 3, `block-height` is removed to avoid confusion.
                // It is replaced with two new keywords: `stacks-block-height` and `tenure-height`.
```

> Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?
 Yes, you can.

Reply:
I updated to the above comment in 7a63f899e8945a770642517e96324330597dbe87.

Reply:
Thanks for the updated explanation.  But, I'm having a hard time convincing myself that the call to `.get_tenure_height()` can _never_ return an `InterpreterError::Expects(..)` due to the tenure height key not being mapped in the MARF tip of the calling context.  For example, what happens if I do the following?

```clarity
;; Can I call this in Nakamoto?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO tenure-height)`
```

What happens with traits defined in Nakamoto but that get called in a pre-Nakamoto context?

```clarity
;; Suppose .whoops-trait is defined in the Nakamoto epoch, and suppose it has this function
;; definition:
;; (define-public (get-tenure-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `tenure-getter`'s `get-tenure-height` function simply returns `tenure-height`.
;; How would this work?
(define-read-only (whoops (tenure-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? tenure-getter get-tenure-height))
```

Reply:
I have similar questions about `block-height`.  Can these work?

```clarity
;; Can I access `block-height` within an `at-block` closure within Nakamoto as long as the block
;; predates Nakamoto, or is this now forbidden?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO block-height)
```

Again, what about calls to a trait implementation?

```clarity
;; Suppose .whoops-trait is defined in epoch 2.x, and suppose it has this function
;; definition:
;; (define-public (get-block-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `block-height-getter`'s `get-block-height` function simply returns `block-height`.
;; How would this work when called in Nakamoto?
(define-public (whoops (block-height-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? block-height-getter get-block-height)))
```